warehouse/client/warehouse.go

```diff
@@ -0,0 +1,141 @@
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/rudderlabs/rudder-server/utils/httputil"
+)
+
+const defaultTimeout = 10 * time.Second
+
+// StagingFile contains the require metadata to process a staging file.
+type StagingFile struct {
+	WorkspaceID   string
+	SourceID      string
+	DestinationID string
+	Location      string
+
+	Schema map[string]map[string]interface{}
+
+	FirstEventAt          string
+	LastEventAt           string
+	TotalEvents           int
+	TotalBytes            int
+	UseRudderStorage      bool
+	DestinationRevisionID string
+	// cloud sources specific info
+	SourceBatchID   string
+	SourceTaskID    string
+	SourceTaskRunID string
+	SourceJobID     string
+	SourceJobRunID  string
+	TimeWindow      time.Time
+}
+
+// legacyPayload is used to maintain backwards compatibility with the /v1 endpoint.
+type legacyPayload struct {
+	WorkspaceID      string
+	Schema           map[string]map[string]interface{}
+	BatchDestination stagingFileBatchDestination
+
+	Location              string
+	FirstEventAt          string
+	LastEventAt           string
+	TotalEvents           int
+	TotalBytes            int
+	UseRudderStorage      bool
+	DestinationRevisionID string
+	// cloud sources specific info
+	SourceBatchID   string
+	SourceTaskID    string
+	SourceTaskRunID string
+	SourceJobID     string
+	SourceJobRunID  string
+	TimeWindow      time.Time
+}
+
+type stagingFileBatchDestination struct {
+	Source      struct{ ID string }
+	Destination struct{ ID string }
+}
+
+type Warehouse struct {
+	baseURL string
+	client  http.Client
```

Comment:
```suggestion
	client  *http.Client
```
Any specific reason we didn't use a pointer?